Address Russ Housley review on MACAddress otherName (#204)
thomas-fossati/draft-tls13-iot: device-serial.md

# Device Identifiers in X.509 Certificates

This note compares the different approaches for encoding device serial
numbers, EUI-48 identifiers, and EUI-64 identifiers in X.509 certificates. It
also explains the relationship between RFC 5280, RFC 7925, RFC 8995, and
`draft-ietf-lamps-macaddress-on`.

## Three Different Kinds of Serial Number

The term "serial number" is overloaded. The following values need to be
distinguished:

1. **Certificate serial number**

   This is the `serialNumber` field in `TBSCertificate`, defined in
   Section 4.1.2.2 of RFC 5280 as `CertificateSerialNumber`. It is assigned by
   the CA and, together with the issuer name, uniquely identifies a
   certificate. It identifies the certificate, not the device.

2. **Device or manufacturer serial number**

   This is an identifier assigned to a device by its manufacturer. When it is
   placed in the certificate subject distinguished name, RFC 5280 provides the
   `X520SerialNumber` attribute:

   ```
   id-at-serialNumber   OBJECT IDENTIFIER ::= { id-at 5 }
   X520SerialNumber    ::= PrintableString
   ```

>    This value is part of the `subject` DN. RFC 5280 defines the attribute and
>    its syntax, but it does not require every device certificate to contain it,
>    nor does it define an EUI-48 or EUI-64 as its required value.

Reviewer: ```suggestion
   This value is part of the `subject` DN. RFC 5280 defines the attribute and
   its syntax, but it does not require every device certificate to contain it,
   nor does it define an EUI-48 or EUI-64 as its required value.
   {{IEEE-802.1AR}} mandates that the SubjectDN not be null (Section 8.6), encouraging the use of the    X520SerialNumber is the primary name for the device.
```

  attribute.
- The `subjectAltName` extension can contain different `GeneralName` forms,
  including `otherName`. An `otherName` is identified by an OID and can carry
  a separately defined ASN.1 value.

RFC 5280 therefore supplies the generic mechanisms, but it does not establish
a universal mapping between a device serial number, an EUI-48, an EUI-64, and
one of these certificate fields.

## RFC 7925: EUI-64 as the Client Identifier

Section 4.4.2 of RFC 7925 requires the identifier in a client certificate to be
an EUI-64. It permits that identifier to appear in either:

- the `subjectAltName`; or
- the leftmost `commonName` component of the subject DN.

RFC 7925 does not specify which `GeneralName` form is used in the
`subjectAltName`. It also does not use or require `X520SerialNumber` for this
purpose.

This leaves two problems:

- An EUI-64 is not a DNS name, URI, IP address, or one of the other
  pre-existing, naturally applicable `GeneralName` forms.
- Encoding the EUI-64 in `commonName` gives a link-layer identifier the
  semantics of a generic DN string and conflicts with the modern practice of
  not using `commonName` for service identity.

> The current TLS/DTLS 1.3 IoT profile lifts the RFC 7925 requirement that every

Reviewer: ```suggestion
This TLS/DTLS 1.3 IoT profile lifts the RFC 7925 requirement that every
```

end-entity identifier be an EUI-64. It nevertheless states that, when an
EUI-64 is used, it must be encoded in the subject DN as
`X520SerialNumber`. That rule is a profiling decision made by this document;
it is not a requirement inherited from RFC 7925.

## RFC 8995: Manufacturer Device Serial Number

Section 2.3.1 of RFC 8995 uses a device serial number to identify a BRSKI
pledge. It explicitly distinguishes this value from the certificate serial

> number.

Reviewer: ```suggestion
Section 2.3.1 of RFC 8995 uses a device serial number to identify a BRSKI
pledge. 
```

The serialNumber != X520SerialNumber distinction seems too many times repeated here.
Do we need the second sentence again?

RFC 8995 identifies the device serial-number field as the

> `X520SerialNumber` attribute defined in Appendix A.1 of RFC 5280. The registrar

Reviewer: ```suggestion
Consistent with {{IEEE-802.1AR}}, RFC 8995 identifies the device serial-number field as the
`X520SerialNumber` attribute defined in Appendix A.1 of RFC 5280. The registrar
```

extracts this certified device serial number from the pledge's IDevID and uses
it in voucher processing.

The important semantic point is that BRSKI needs a stable manufacturer device
identifier. RFC 8995 does not require this value to be an EUI-48 or EUI-64.
For example, it can be a manufacturer-assigned value such as `WI-3005`.

The BRSKI approach is therefore:

- **Meaning:** manufacturer or device serial number
- **Location:** subject distinguished name
- **Attribute:** `X520SerialNumber`
- **Encoding:** `PrintableString`
- **Primary use:** stable device identity and correlation with manufacturing
  and voucher records

## MACAddress otherName: Typed EUI-48 and EUI-64 Values

`draft-ietf-lamps-macaddress-on` defines a dedicated
`GeneralName.otherName` form named `MACAddress`. Its OID is:

```
id-on-MACAddress OBJECT IDENTIFIER ::= { 1 3 6 1 5 5 7 8 12 }
```

> 

Reviewer: ```suggestion
{{!I-D.ietf-lamps-macaddress-on}} defines a dedicated
`GeneralName.otherName` form named `MACAddress`. 

```

It seems more DRY to leave the OID to the other RFC.  That document is ahead of us (in the RFC Q), so a normative reference won't hurt us.

Neither encoding universally replaces the other because they represent
different kinds of identity:

- Use `X520SerialNumber` when the certificate carries the manufacturer's
  stable device serial number, including when that serial number happens to
  use an EUI-64-shaped string.
- Use `MACAddress otherName` when the certificate asserts that a particular
  EUI-48 or EUI-64 belongs to the subject device or one of its interfaces.

If the same EUI-64 is deliberately used for both purposes, a profile must
decide whether it is encoded once, encoded in both fields, or treated as only
one of the two semantic identifiers. Encoding the value in both fields creates
duplication and requires rules for handling mismatches.

## Consequences for the TLS/DTLS IoT Profile

The current statement that every EUI-64 identifier must be encoded as
`X520SerialNumber` excludes the typed `MACAddress otherName` representation.
It also conflates two cases:

1. an EUI-64 used as a manufacturer's stable device serial number; and
2. an EUI-64 used as a link-layer interface identifier.

The profile should permit the representation appropriate to the intended

> semantics. One possible formulation is:

Reviewer: Like, are we saying this, or not?  Where would this formulation go, if not here?

> A manufacturer-assigned device serial number included in the Subject DN
> MUST be encoded in the X520SerialNumber attribute. If an EUI-48 or EUI-64 is

> > included as a link-layer identifier, it SHOULD be encoded in the

Reviewer: ```suggestion
> included is a link-layer identifier, it SHOULD also be encoded in the
```

> subjectAltName using the MACAddress otherName defined in
> draft-ietf-lamps-macaddress-on. An EUI-64 that serves as the
> manufacturer-assigned device serial number MAY instead be encoded in the
> X520SerialNumber attribute.

The final normative language should also decide:

> - whether `MACAddress otherName` is a `MUST`, `SHOULD`, or `MAY`;

Reviewer: SHOULD if EUI-xx.
X520SerialNumber needs to remain to be 802.1AR compatible, in the sense that SubjectDN is otherwise... empty, which 1AR does not permit.   Loading a certificate to a picky end device (like, via PKCS12, yes, ick), that has a null DN likely fails, but adding otherName should be benign.

Reviewer 2: I agree that 802.1.AR compatible is good to have.

- whether the legacy `X520SerialNumber` encoding of an EUI-64 remains
  permitted for compatibility;
- whether the same identifier may appear in both fields;
- what a relying party does if both fields are present but differ;
- whether certificates may contain several MACAddress values for
  multi-interface devices; and
- whether interface replacement, MAC randomization, or locally administered
  addresses are within scope.

## Privacy and Lifecycle Considerations

Both approaches can expose stable identifiers to certificate recipients.
TLS 1.3 encrypts certificates during the handshake against passive on-path

> observers, but the peer still learns the identifier.

Reviewer: I think that a active on-path attacker does not see the client certificate, because such a thing fails the TLS1.3 handshake, right?

Author: This was an issue that came up in one of the reviews. The point is that there are cases where an active attack can be successful, for example in a device-to-device communication pattern where the attacker pretends to be the other device. To be precise here one would have to expand the entire story a little bit.

Reviewer 2: at least for the server side.

An EUI-48 or EUI-64 can reveal organizational allocation information and can
enable correlation across networks or application contexts. A stable device
serial number has similar correlation risks. Certificates should therefore
contain these identifiers only when they are required for authentication or

> authorization.

Reviewer: ```suggestion
An EUI-48 or EUI-64 can reveal organizational allocation information and can
enable correlation across networks or application contexts. 
A stable device serial number has similar correlation risks. 
Environments that are concerned about such traffic analysis SHOULD use an enrollment protocol
to migrate from identifiable IDevID Certificates to less identifiable (operational) LDevID certificates.
```

Reviewer: (I don't think we should encourage IDevID to avoid EUI-xx, rather to limit their use)

The CA also needs to validate the identifier's relationship to the subject.
For a MACAddress value, the LAMPS draft requires the CA to ensure that the
address is owned by, or expected to be owned by, the subject device for the
certificate lifetime. This requirement can be difficult for replaceable
interfaces, virtual interfaces, locally administered addresses, and MAC
address randomization.

> ## References

Reviewer: Patch the references for this?  Shall I do that...

Author: We will patch the references once we put the text into the draft itself
